lib/internal/process/coverage.js

```diff
@@ -0,0 +1,69 @@
+'use strict';
+const path = require('path');
+const { mkdirSync, writeFileSync } = require('fs');
+const inspector = require('inspector');
+// TODO(addaleax): add support for coverage to worker threads.
+const hasInspector = process.config.variables.v8_enable_inspector === 1 &&
+  require('internal/worker').isMainThread;
+
+let session;
+
+function writeCoverage() {
+  if (!hasInspector || !session) {
+    return;
+  }
+
+  const filename = `coverage-${process.pid}-${Date.now()}.json`;
+  try {
+    // TODO(bcoe): switch to mkdirp once #22302 is addressed.
+    mkdirSync(process.env.NODE_V8_COVERAGE);
+  } catch (err) {
+    if (err.code !== 'EEXIST') {
+      console.error(err);
+      return;
+    }
+  }
+
+  const target = path.join(process.env.NODE_V8_COVERAGE, filename);
+
+  try {
+    session.post('Profiler.takePreciseCoverage', (err, coverageInfo) => {
+      try {
+        if (err) throw err;
+        writeFileSync(target, JSON.stringify(coverageInfo));
+      } catch (err) {
+        console.error(err);
+      }
+    });
+  } catch (err) {
+    console.error(err);
+  } finally {
+    session.disconnect();
+    session = null;
+  }
+}
+
+exports.writeCoverage = writeCoverage;
+
+function setup() {
+  if (!hasInspector) {
+    return;
+  }
+
+  session = new inspector.Session();
+  session.connect();
+  session.post('Profiler.enable');
+  session.post('Profiler.startPreciseCoverage', { callCount: true,
+                                                  detailed: true });
```

Comment:
Do we want to also expose a way to use other coverage modes? This primarily has performance implications. E.g. `callCount: false` would allow a function to be inlined after it has been covered. `detailed: true` would not require coverage-related bytecode to be emitted.

Reply:
I'd argue for simplicity, exposing only the most detailed mode, and making sure it is as fast as possible. `callCount: true / detailed: true` should also allow inlining (it does require special bytecode emission though).

Reply:
I'm with @schuay on this one, I'd lean towards starting simple. I can imagine following up with the option for more complex configuration.